diff --git a/wamr-compiler/CMakeLists.txt b/wamr-compiler/CMakeLists.txt
index fa11192f..47652586 100644
--- a/wamr-compiler/CMakeLists.txt
+++ b/wamr-compiler/CMakeLists.txt
@@ -181,7 +181,9 @@ include (${SHARED_DIR}/utils/shared_utils.cmake)
 include (${SHARED_DIR}/utils/uncommon/shared_uncommon.cmake)
 include (${IWASM_DIR}/libraries/thread-mgr/thread_mgr.cmake)
 include (${IWASM_DIR}/libraries/libc-builtin/libc_builtin.cmake)
-include (${IWASM_DIR}/libraries/libc-wasi/libc_wasi.cmake)
+if (NOT MINGW)
+  include (${IWASM_DIR}/libraries/libc-wasi/libc_wasi.cmake)
+endif()
 include (${IWASM_DIR}/libraries/lib-pthread/lib_pthread.cmake)
 include (${IWASM_DIR}/common/iwasm_common.cmake)
 include (${IWASM_DIR}/interpreter/iwasm_interp.cmake)
@@ -253,9 +255,11 @@ add_executable (wamrc main.c)
 
 if (NOT MSVC)
   target_link_libraries (wamrc aotclib vmlib LLVMDemangle ${LLVM_AVAILABLE_LIBS} ${lib_ubsan}
-                         -lm -ldl -lpthread ${lib_lldb})
+                         -lm -lpthread ${lib_lldb})
   if (MINGW)
       target_link_libraries (wamrc -lssp -lWs2_32)
+  else()
+      target_link_libraries (wamrc aotclib vmlib LLVMDemangle -ldl)
   endif()
 else()
   target_link_libraries (wamrc aotclib vmlib  ${lib_lldb} ${LLVM_AVAILABLE_LIBS} ${lib_ubsan})
